firmware/components/wifi/include/wifi_provisioning.h

```diff
@@ -0,0 +1,206 @@
+ */
+
+#pragma once
+
+#include <stdio.h>
+#include <string.h>
+
+#include <freertos/FreeRTOS.h>
+#include <freertos/task.h>
+#include <freertos/event_groups.h>
+#include <queue.h>
+
+#include <esp_log.h>
+#include <esp_wifi.h>
+#include <esp_event.h>
+#include <nvs_flash.h>
+
+#include <wifi_provisioning/manager.h>
+#include <wifi_provisioning/scheme_softap.h>
+
+namespace shrapnel {
+namespace wifi_provisioning {
+
+class WiFiProvisioning final {
+private:
+
+    enum event_type_t {
+        EVENT_PROVISIONING_SUCCESSFUL,
+        EVENT_PROVISIONING_FAILED,
+    };
+
+    struct Event {
+        event_type_t type;
+    };
+
+    static constexpr char TAG[5] = "wifi";
+
+    Queue<Event> queue;
+
+static void event_handler(void* arg, esp_event_base_t event_base,
+                          int32_t event_id, void* event_data)
+{
+    assert(arg != nullptr);
+
+    auto queue = static_cast<Queue<Event> *>(arg);
+
+    if (event_base == WIFI_PROV_EVENT) {
+        switch (event_id) {
+            case WIFI_PROV_START:
+                ESP_LOGI(TAG, "Provisioning started");
+                break;
+            case WIFI_PROV_CRED_RECV: {
+                wifi_sta_config_t *wifi_sta_cfg = (wifi_sta_config_t *)event_data;
+                ESP_LOGI(TAG, "Received Wi-Fi credentials"
+                         "\n\tSSID     : %s\n\tPassword : %s",
+                         (const char *) wifi_sta_cfg->ssid,
+                         (const char *) wifi_sta_cfg->password);
+                break;
+            }
+            case WIFI_PROV_CRED_FAIL: {
+                wifi_prov_sta_fail_reason_t *reason = (wifi_prov_sta_fail_reason_t *)event_data;
+                ESP_LOGE(TAG, "Provisioning failed!\n\tReason : %s",
+                         (*reason == WIFI_PROV_STA_AUTH_ERROR) ?
+                         "Wi-Fi station authentication failed" : "Wi-Fi access-point not found");
+
+                Event event{.type = EVENT_PROVISIONING_FAILED};
+                int rc = queue->send(&event, 100 / portTICK_PERIOD_MS);
+                if(rc != true)
+                {
+                    ESP_LOGE(TAG, "failed to send to queue");
+                }
+                break;
+            }
+            case WIFI_PROV_CRED_SUCCESS: {
+                ESP_LOGI(TAG, "Provisioning successful");
+
+                Event event{.type = EVENT_PROVISIONING_SUCCESSFUL};
+                int rc = queue->send(&event, 100 / portTICK_PERIOD_MS);
+                if(rc != true)
+                {
+                    ESP_LOGE(TAG, "failed to send to queue");
+                }
+                break;
+            }
+            case WIFI_PROV_END:
+                ESP_LOGI(TAG, "Provisioning end");
+                break;
+            default:
+                break;
+        }
+    } else if (event_base == WIFI_EVENT && event_id == WIFI_EVENT_STA_START) {
+        // TODO this is causing an error when not provisioned. Is it doing
```

Comment:
Check if this is called when credentials are provided during provisioning or when provisioning is already complete.

Maybe it can be removed?